### PR TITLE
2d plot titles

### PR DESCRIPTION
**Pull request by clarkfitzg**
Working on faceted plotting now. This will be helpful. It adds a title if the array came from a slice, and labels the colorbar with the variable name.

image

### CODE REVIEW

File: xray/plot/plot.py [hunk 1]
```diff
@@ -55,6 +55,17 @@ def _load_default_cmap(fname='default_colormap.csv'):
...
 
 
+def _title_for_slice(darray):
+    '''
+    If the dataarray comes from a slice we can show that info in the title
+    '''
+    title = []
+    for key, value in darray.coords.items():
+        if value.size == 1:
+            title.append('{} = {}'.format(key, value.values))
```
Comment (jhamman) [inline xray/plot/plot.py]: don't you want to use integer placeholders for python 2.6? 
```python
title.append('{0} = {1}'.format(key, value.values))
```

File: xray/plot/plot.py [hunk 2]
```diff
@@ -55,6 +55,17 @@ def _load_default_cmap(fname='default_colormap.csv'):
     return LinearSegmentedColormap.from_list('viridis', cm_data)
 
 
+def _title_for_slice(darray):
+    '''
+    If the dataarray comes from a slice we can show that info in the title
+    '''
+    title = []
+    for key, value in darray.coords.items():
```
Comment (jhamman) [inline xray/plot/plot.py]: will this catch any size one coordinate, even if it is not a slice?

File: xray/plot/plot.py [hunk 3]
```diff
@@ -55,6 +55,17 @@ def _load_default_cmap(fname='default_colormap.csv'):
...
+        if value.size == 1:
```
Comment (jhamman) [inline xray/plot/plot.py]: will this catch any size one coordinate, even if it is not a slice?

Comment (clarkfitzg) [inline xray/plot/plot.py]: Yes. Here's a 4d array with a single size 1 coordinate:
```
  * dim_0    (dim_0) int64 0 1 2
  * dim_1    (dim_1) int64 0 1 2 3
  * dim_2    (dim_2) int64 0 1 2 3 4 5 6 7 8 9
  * dim_3    (dim_3) int64 0

In [11]: _title_for_slice(b)
Out[11]: 'dim_3 = [0]'

In [13]: _title_for_slice(b.isel(dim_1=2))
Out[13]: 'dim_1 = 2, dim_3 = [0]'
```
It shows square brackets around the value, since it's an array in this case.

However, this won't come up in the current implementation since we don't allow 2d plots to work on say a 2 x 4 x 1 array:

```
In [14]: b.isel(dim_1=2).plot.contourf()

...

ValueError: contourf plots are for 2 dimensional DataArrays. Passed DataArray has 3 dimensions
```

I've thought about making this a special case and allowing it to plot. Thoughts?

Comment (jhamman) [inline xray/plot/plot.py]: Yeah, why don't you grab the titles, then squeeze the DataArray? If it can't be squeezed to a 2d array, then raise that error.

**Comment by clarkfitzg** [comment]
@jhamman good idea :+1:

**Comment by jhamman** [comment]
@clarkfitzg - in your eyes, is this done? If so, I'll give it another look?

**Comment by clarkfitzg** [comment]
@jhamman Yes, ready for review.

File: xray/plot/plot.py [hunk 4]
```diff
@@ -55,6 +55,17 @@ def _load_default_cmap(fname='default_colormap.csv'):
     return LinearSegmentedColormap.from_list('viridis', cm_data)
 
 
+def _title_for_slice(darray):
+    '''
+    If the dataarray comes from a slice we can show that info in the title
+    '''
+    title = []
+    for dim, coord in darray.coords.items():
```
Comment (jhamman) [inline xray/plot/plot.py]: use `iteritems` from `core.pycompat`:

```python
for dim, coord in iteritems(darray.coords):
    ...
```

File: xray/plot/plot.py [hunk 5]
```diff
@@ -456,11 +472,16 @@ def newplotfunc(darray, ax=None, xincrease=None, yincrease=None,
...
+        # Label the plot with metadata
+        if add_labels:
+            ax.set_xlabel(xlab)
+            ax.set_ylabel(ylab)
+            ax.set_title(_title_for_slice(darray))
 
         if add_colorbar:
-            plt.colorbar(primitive, ax=ax, extend=cmap_params['extend'])
+            cbar = plt.colorbar(primitive, ax=ax, extend=cmap_params['extend'])
+            if darray.name and add_labels:
```
Comment (jhamman) [inline xray/plot/plot.py]: since `DataArray.__init__()` defines the `name` attribute as `None` by default, probably better to use the `is not None` syntax:
```python
if darray.name is not None and add_labels:```

File: xray/test/test_plot.py [hunk 6]
```diff
@@ -375,6 +375,28 @@ def test_diverging_color_limits(self):
...
+        self.assertIn(self.darray.name, alltxt)
+
+    def test_no_labels(self):
+        self.darray.name = 'testvar'
+        self.plotmethod(add_labels=False)
+        alltxt = [t.get_text() for t in plt.gcf().findobj(mpl.text.Text)]
+        alltxt = set(alltxt)
+        for string in ['x', 'y', 'testvar']:
+            self.assertNotIn(string, alltxt)
+
```
Comment (jhamman) [inline xray/test/test_plot.py]: Since we added the `squeeze` functionality in this PR, let's add a test to make sure we can handle plotting data that is shaped in an edge case.  For example if `da.shape == (5, 1)`, I think we'll end up in the plot2d and will fail after the squeeze.

File: xray/plot/plot.py (hunk 5 above)
Comment (clarkfitzg) [inline xray/plot/plot.py]: Not sure I follow. As it stands this works for `None` or an empty string-

```
In [10]: a = xray.DataArray([1, 2, 3])

In [11]: if a.name:
   ....:     print('fail')
   ....:

In [12]: a.name = ''

In [13]: if a.name:
    print('fail')
   ....:

In [14]: a.name = 'somename'

In [15]: if a.name:
   ....:     print('success')
   ....:
success

In [16]:
```

File: xray/test/test_plot.py (hunk 6 above)
Comment (clarkfitzg) [inline xray/test/test_plot.py]: nice catch!

**Comment by clarkfitzg** [comment]
Will finish this up tomorrow morning.

File: xray/plot/plot.py (hunk 5 above)
Comment (jhamman) [inline xray/plot/plot.py]: This was really a nitpicky comment. The only sure way to know the user hasn't set the name attribute to something is to compare it to `None`.  For example, `False` and `0` both evaluate to `False` so if the user, for some reason, set their DataArray to one of those, the current syntax would fail to add the label: 

```python
In [9]: a.name = 0

In [10]: if a.name:
   ....:     print('fail')

In [11]:
```

Maybe not the worst thing, but *could* lead to some confusion at some point.

Comment (clarkfitzg) [inline xray/plot/plot.py]: Makes sense. Just discovered this works: `a.name = (1, 2, 3)` but then fails to print. The `name` attribute is used to give the correct name with `to_dataset`. Do you know if it's used for other purposes also?

If I set an integer as a data variable name then the dataset won't save to netcdf.
So maybe it's worth restricting the name to be a string?

Comment (jhamman) [inline xray/plot/plot.py]: I think it makes sense to restrict `a.name` to a string. Probably a separate PR though.

File: xray/plot/plot.py [hunk 7]
```diff
@@ -469,14 +481,21 @@ def newplotfunc(darray, ax=None, xincrease=None, yincrease=None,
...
     def plotmethod(_PlotMethods_obj, ax=None, xincrease=None, yincrease=None,
-                   add_colorbar=True, vmin=None, vmax=None, cmap=None,
+                   add_colorbar=True, add_labels=True, vmin=None, vmax=None, cmap=None,
                    center=None, robust=False, extend=None, levels=None,
                    **kwargs):
-        return newplotfunc(_PlotMethods_obj._da, ax=ax, xincrease=xincrease,
-                           yincrease=yincrease, add_colorbar=add_colorbar,
-                           vmin=vmin, vmax=vmax, cmap=cmap,
-                           center=center, robust=robust, extend=extend,
-                           levels=levels, **kwargs)
```
Comment (clarkfitzg) [inline xray/plot/plot.py]: Changes here make it easier to keep the function / methods consistent.

**Comment by shoyer** [comment]
For interactive use, I think plotting squeezed data is probably what users want. But I worry that this will make the result of plotting functions less predictable. Maybe we should restrict squeezing the data to only the magic `DataArray.plot()` method?

For axis titles, what happens if the data array has many scalar coordinates? Scalar coordinates can be directly assigned, not just created from indexing. My sense is that labelling facets will be more reliable if it is done explicitly in the same code that is doing the indexing/groupby.

**Comment by clarkfitzg** [comment]
> Maybe we should restrict squeezing the data to only the magic DataArray.plot() method?

Works for me.  The error message should be informative enough in the other cases.

Scalar coordinates- this is the current behavior:
```
In [16]: a = xray.DataArray(np.random.randn(3, 4))

In [17]: a.coords['foo'] = 123

In [18]: a.plot()
```
image

So scalar coordinates do show up in the title, which is what I was going for. I could add a test for this.

> labelling facets will be more reliable if it is done explicitly in the same code that is doing the indexing/groupby.

That's the approach that I've moved to after spending more time on it. This PR adds the argument `add_labels` to 2d plotting functions. Now for the faceting I'll actually use `add_labels=False`.

**Comment by shoyer** [comment]
OK, seems pretty reasonable on the title then. Just note that some datasets (especially from model output) contain *a lot* of scalar variables, so that could easily get unwieldy. We might need to revisit this once it gets more use in the wild. For example, truncating the title at some reasonable line length might be a good idea.

File: xray/core/dataarray.py [hunk 8]
```diff
@@ -1091,6 +1091,16 @@ def plot(self):
         '''
         return _PlotMethods(self)
 
+    def _title_for_slice(self):
```
Comment (shoyer) [inline xray/core/dataarray.py]: To me, this logic seems more appropriate for a helper function in the plotting module (closer to where it is called).

Comment (clarkfitzg) [inline xray/core/dataarray.py]: Was waffling around where to put it. It's been in 3 different places over the course of this PR- here, in utils.py, and in the plot module. If you think the plot module is the best place then I'll put it there.

Comment (shoyer) [inline xray/core/dataarray.py]: This is fine -- it doesn't matter very much

**Comment by shoyer** [comment]
feel free to merge
